res/controllers/lib/mididispatcher.mjs

```diff
@@ -0,0 +1,24 @@
+export class MidiDispatcher {
+    constructor(noteOff) {
+        this.noteOff = noteOff;
+        this.inputMap = new Map();
+    }
+    registerInputCallback(midiBytes, callback) {
+        // JavaScript is broken and believes [1,2] === [1,2] is false, so
+        // JSONify the Array to make it usable as a Map key.
+        const key = JSON.stringify(midiBytes);
+        this.inputMap.set(key, callback);
+        if (this.noteOff === true && ((midiBytes[0] & 0xF0) === 0x90)) {
+            const noteOffBytes = [midiBytes[0] - 0x10, midiBytes[1]];
+            const noteOffKey = JSON.stringify(noteOffBytes);
+            this.inputMap.set(noteOffKey, callback);
+        }
+    }
+    receiveData(data, timestamp) {
+        const key = JSON.stringify([data[0], data[1]]);
```

Comment:
Hmm this assumes that MIDI data is at least 2 bytes, so it can't be used to map things like `FA` (sent by my controller when pressing the play button of the built-in drum computer).

Reply:
This class doesn't need to handle every single possibility for esoteric MIDI messages. The controller module's `receiveData` function could check for exceptional messages and return before calling `MidiDispatcher.receiveData`.

Reply:
According to [this list](https://www.midi.org/specifications-old/item/table-2-expanded-messages-list-status-bytes), all messages are 3 bytes, except:
1. Messages starting with C or D are 2 bytes long where byte 2 is the value
2. Messages starting with F are 1 byte long or "special messages" like SysEx

I think we only need 2 cases:
1. First midi nibble is C, D or F: only use first byte as key
2. Use first 2 bytes in all other cases

Reply:
I added support for 0xC (program change) and 0xD (aftertouch) messages, but I don't think it would be helpful to add logic for handling 0xF messages which are highly variable. I suggest implementing controller-specific logic for handling those messages in the controller module's `receiveData` function and returning before calling `MidiDispatcher.receiveData`.

Reply:
FWIW there is a single controller mapping in Mixxx (Akai LPD8) that uses 0xC (program change) MIDI messages and zero that use 0xD (aftertouch).